### PR TITLE
feat(surveys): `posthog` changes to support valid HTML question rendering

### LINKED ISSUE #22509: Improve HTML markup experience in the survey editor
## Bug description

Right now, the HTML question editing experience for the survey editor doesn't work quite right.  The main issue is that the "Text" box field is evaluating HTML tags instead of just rendering them.  This leads to 2 problems:

1. Users can create HTML questions without paying for the HTML questions feature, since they can just put HTML styling into the Text box
2. The text box shouldn't evaluate HTML at all, it should render tags as strings (this is my opinion, it's not necessarily a _bug_ but it feels inaccurate to me that you can put HTML into a Text box and it still works.

## How to reproduce

1. Create a new (or edit an existing) survey
2. Go to the question editor
3. Put HTML into the "Text" field. 
4. Notice in the popup survey preview that the HTML is evaluated

### PR DESCRIPTION
**Pull request by dmarticus**
## Summary

I modified the survey renderer to differentiate between `Text` and `HTML` references when writing question descriptions.  Before this change, we had a bug where users could bypass the HTML question description markup paywall by writing raw HTML into their Text description (and it would still be interpolated as HTML if the tags were valid), so this change makes it so that we only interpolate HTML in the popup survey if the HTML feature is (a) paid for and (b) enabled.

### Important Note

This PR includes changes to the `Survey` Model, which is a Core model that is serialized throughout our app and stored in our Postgres tables.  Specifically, I modified the `BasicSurveyQuestion` and `SurveyAppearance` models to store the content type of the `description fields.  

Schemas before
```
export interface SurveyQuestionBase {
    question: string
    description?: string | null
    optional?: boolean
    buttonText?: string
}

export interface SurveyAppearance {
    backgroundColor?: string
    submitButtonColor?: string
    // TODO: remove submitButtonText in favor of buttonText once it's more deprecated
    submitButtonText?: string
    ratingButtonColor?: string
    ratingButtonActiveColor?: string
    borderColor?: string
    placeholder?: string
    whiteLabel?: boolean
    displayThankYouMessage?: boolean
    thankYouMessageHeader?: string
    thankYouMessageDescription?: string
    autoDisappear?: boolean
    position?: string
    // widget only
    widgetType?: 'button' | 'tab' | 'selector'
    widgetSelector?: string
    widgetLabel?: string
    widgetColor?: string
    shuffleQuestions?: boolean
}
```

Schemas after
```
export interface SurveyQuestionBase {
    question: string
    description?: string | null
    descriptionContentType?: SurveyQuestionDescriptionContentType // new field
    optional?: boolean
    buttonText?: string
}

export interface SurveyAppearance {
    backgroundColor?: string
    submitButtonColor?: string
    // TODO: remove submitButtonText in favor of buttonText once it's more deprecated
    submitButtonText?: string
    ratingButtonColor?: string
    ratingButtonActiveColor?: string
    borderColor?: string
    placeholder?: string
    whiteLabel?: boolean
    displayThankYouMessage?: boolean
    thankYouMessageHeader?: string
    thankYouMessageDescription?: string
    thankYouMessageDescriptionContentType?: SurveyQuestionDescriptionContentType // new field
    autoDisappear?: boolean
    position?: string
    // widget only
    widgetType?: 'button' | 'tab' | 'selector'
    widgetSelector?: string
    widgetLabel?: string
    widgetColor?: string
    shuffleQuestions?: boolean
}
```

Given that we now have new fields that posthog-js requires to differentiate between how to render descriptions, we now need some strategy to backfill this data so that old surveys can render descriptions appropriately.  Fortunately, only one (1) user has the question description field even populated (and they're using plain text), and no one uses this field for the confirmation message description, so this is pretty safe.  Furthermore, in my `posthog-js` changes, [I elected to default to `Text` rendering](https://github.com/PostHog/posthog-js/pull/1206/files#diff-ca75b1ebdb1e4a75293f5a5f4d0bdc7b94c5ca7243b5ac9ebb49147a6dd34d83R229-R231).

In summary, this feature isn't really used very much (yet), which is good because this makes a change like this pretty safe.

### More info

Needs to ship with the associated `posthog-js` changes here: https://github.com/PostHog/posthog-js/pull/1206/

Closes: https://github.com/PostHog/posthog/issues/22509

## Changes

See the ticket for how the HTML parsing behavior used to look.

New behavior looks like this: 

## Does this work well for both Cloud and self-hosted?

Yes

### Manual Test Cases

(Ran these tests for every question field and for the thank you message description)
- Text renders correctly
- HTML renders correctly
- Switching between text and HTML causes the tags to be manually printed (if text) and interpolated (if HTML)
-

### CODE REVIEW

**Comment by neilkakkar** [comment]
Heya! How did you figure out there's only one user using html descriptions? This doesn't sound right, I'm pretty sure there's multiple just in our team, and I'm _very_ sure I've seen multiple customers with html surveys during support.

One other thing worth noting is that this same problem applies not just to descriptions, but the question text, and button text as well - they're all rendered as HTML always.

I love what you're doing here, I think we should extend this to a global survey render type setting: If enabled, and set, we render all fields as HTML. If not, we render all fields as text. I.e. it's not just for description.

---

For backwards compatibility, and not breaking existing running surveys, I'd recommend defaulting to HTML for all surveys where this is not set. Plus, for all new surveys, we always make sure to set it.

Oh, and also, we'd want API level validation here - if `html` isn't allowed, anyone using the API to create a new survey also shouldn't be allowed to set `html` as the render type

**Comment by dmarticus** [comment]
@neilkakkar, you're absolutely right, there was a bug in my postgres syntax.  I amended it today and found that there were [101 surveys that had either HTML](http://metabase-prod-us/question/681-surveys-using-html-descriptions-in-either-the-questions-or-confirmation-messages) in the question description or confirmation message description (in case you can't load my metabase dashboard, here's the query)
```sql
SELECT 
    COUNT(*) 
FROM (
    SELECT 
        id
    FROM 
        posthog_survey,
        jsonb_array_elements(questions) AS elem
    WHERE 
        elem->>'description' IS NOT NULL
        AND elem->>'description' <> ''
        AND elem->>'description' ~ '<[^>]+>'
    UNION
    SELECT 
        id
    FROM 
        posthog_survey
    WHERE 
        appearance ? 'thankYouMessageDescription'
        AND appearance->>'thankYouMessageDescription' IS NOT NULL
        AND appearance->>'thankYouMessageDescription' <> ''
        AND appearance->>'thankYouMessageDescription' ~ '<[^>]+>'
) AS combined_results;
```
and [49 in EU](http://metabase-eu/question/157-surveys-using-html-descriptions-in-either-the-questions-or-confirmation-messages), for a total of 150 surveys.  So we'll definitely need to figure out a backfill strategy that doesn't bork all of these existing surveys that are marked up.

To your point on not wanting to break existing surveys, I'm fine defaulting to HTML if the value was never set in the survey itself.  I'll make that change in `posthog-js` (since that's where the rendering actually happens), and then I'll make that new and updated surveys start to save the content type in the survey definition.

**Comment by dmarticus** [comment]
@neilkakkar I made more progress on this today but there's still a few things I'd to discuss with you before I go much further.

- I'm using `useEffect()` in the two components parent components where I use the `HTMLEditor` component ([example](https://github.com/PostHog/posthog/pull/22533/files#diff-a12eb517d1b5fef23ebe3b6933695b3c304f8a4c14c436becb1cbc26604acad5R91-R97)), which I know we shouldn't do. I tried using kea logic to do this instead (here the [action](https://github.com/PostHog/posthog/pull/22533/files#diff-0a051008afcf371e7c0f56a8f251f7d5fd2b20c2c8b491d1b4bd157347e593c9R149-R152) and [reducer](https://github.com/PostHog/posthog/pull/22533/files#diff-0a051008afcf371e7c0f56a8f251f7d5fd2b20c2c8b491d1b4bd157347e593c9R513-R525) what I added in my attempt to do this), but I wasn't able to get changes to persist across clicks; any time I clicked a different tab, it just reverted back to the default.  I'd love to understand what my approach should be here to handle managing state where it involves changing the values of the survey on the page, too.
- I'm not sure exactly where the HTML validation on the backend is happening.  I'm trying to answer this question:

> Oh, and also, we'd want API level validation here - if html isn't allowed, anyone using the API to create a new survey also shouldn't be allowed to set html as the render type

but I don't know where to look for this just yet.  

- Finally, I'm not really sure what sorts of tests I should be writing here, but I feel like this is a significant enough change that I'd like to write them.  Would love to get your opinion on how to approach this.

File: posthog/api/test/test_survey.py [hunk 1]
```diff
@@ -1,28 +1,28 @@
...
     QueryMatchingTest,
-    snapshot_postgres_queries,
-    snapshot_clickhouse_queries,
     _create_event,
+    snapshot_clickhouse_queries,
+    snapshot_postgres_queries,
 )
 
-from posthog.models import FeatureFlag
-
```
Comment (dmarticus) [inline posthog/api/test/test_survey.py:25]: ooof, that was an auto-reordering from my editor.  Should I undo this?

File: posthog/test/base.py [hunk 2]
```diff
@@ -351,14 +348,28 @@ def setUp(self):
         cache.clear()
         TEST_clear_instance_license_cache()
 
-        # Sets the cloud mode to stabilise things tests, especially num query counts
-        # Clear the is_rate_limit lru_Caches so that they does not flap in test snapshots
+        # Sets the cloud mode to stabilize things tests, especially num query counts
+        # Clear the is_rate_limit lru_Caches so that they do not flap in test snapshots
```
Comment (dmarticus) [inline posthog/test/base.py:352]: flyby grammar policing 🚓

File: posthog/test/base.py [hunk 3]
```diff
@@ -351,14 +348,28 @@ def setUp(self):
...
+        return organization
+
+    def create_team_with_organization(self, organization):
+        return Team.objects.create(organization=organization, name="Test Team")
+
+    def create_user_with_organization(self, organization):
+        user = User.objects.create_user(email="testuser@example.com", first_name="Test", password="password")
+        organization.members.add(user)
+        return user
+
```
Comment (dmarticus) [inline posthog/test/base.py:372]: added some functionality to our APIBaseTest class to support adding/removing paid features.

File: posthog/test/base.py (hunk 2 above)
Comment (neilkakkar) [reply posthog/test/base.py:352]: Americans 💪 😅  (but yes thank you our default is American English no matter how much British English I try to sneak in 😂 )

File: posthog/test/base.py (hunk 3 above)
Comment (neilkakkar) [reply posthog/test/base.py:372]: very nice

File: frontend/src/scenes/surveys/SurveyAppearanceUtils.tsx [hunk 4]
```diff
@@ -50,22 +51,36 @@ export function PresentationTypeCard({
...
     value?: string
     onChange: (value: any) => void
-    writingHTMLDescription: boolean
-    setWritingHTMLDescription: (writingHTML: boolean) => void
+    onTabChange: (key: SurveyQuestionDescriptionContentType) => void
+    initialActiveTab: SurveyQuestionDescriptionContentType | undefined
     textPlaceholder?: string
 }): JSX.Element {
     const { surveysHTMLAvailable } = useValues(surveysLogic)
+    const { activeTabKey } = useValues(surveyLogic)
```
[the remark this answers is not in the document]
Comment (neilkakkar) [reply frontend/src/scenes/surveys/SurveyAppearanceUtils.tsx]: Ah yeah this is becoming problematic - it doesn't set the actual value in the survey, but sets it in the UI, making them de-sync 😬

[the remark this answers is not in the document]
Comment (neilkakkar) [reply frontend/src/scenes/surveys/SurveyAppearanceUtils.tsx]: I pushed a proposal, let me know what you think! DIdn't test this with the permissions for html though - so not sure if the flow there works well.

**Review by neilkakkar — COMMENTED** [review]
One comment, keen to hear your thoughts

**Comment by dmarticus** [comment]
I accidentally made a bad merge into this branch, so I'm closing it in favor of https://github.com/PostHog/posthog/pull/22699
